File: scripts/generate-data.js
const fs = require("fs");
const Country = require("../modules/country");
const Report = require("../modules/report");

const countries = [];
const reportTitles = [];

function getCountryByName(name) {
  for (const country of countries) {
    if (country instanceof Country) {
      if (country.name == name) return country;
    }
  }
}

const names = fs
  .readFileSync("resources/data/names.txt")
  .toString()
  .split("\n");
names.forEach((name) => countries.push(new Country(name)));

const flagImageUrls = JSON.parse(
  fs.readFileSync("resources/data/flag-img-urls.json").toString()
);
for (const name in flagImageUrls) {
  getCountryByName(name).flagImageUrl = flagImageUrls[name];
}

fs.readdirSync("resources/data")
  .sort()
  .forEach((filename) => {
    if (filename.endsWith(".csv")) {
      const title = filename.slice(0, -4).split("_").join(" ");
      if (!reportTitles.includes(title)) {
        reportTitles.push(title);
      }
      const file = fs.readFileSync(`resources/data/${filename}`);
      const rows = file.toString().split("\n");
      rows.shift();
      rows.forEach((row) => {
        const tokens = row.split(",");
        const country = getCountryByName(tokens[1]);
        if (country !== undefined) {
          country.reports.push(
            new Report(title, parseInt(tokens[0]), parseFloat(tokens[2]))
          );
        }
      });
    }
  });

fs.writeFileSync(
  "resources/data/data.json",
  JSON.stringify({ countries: countries, reportTitles: reportTitles }, null, 2)
);

File: generate_data.py
import json
import os


countries, report_titles = [], []


def get_country_by_name(name):
    for country in countries:
        if country["name"] == name:
            return country


with open("resources/names.txt") as file:
    names = file.read().split("\n")
    for name in names:
        countries.append({"name": name, "flag": "", "reports": []})
    file.close()

with open("resources/flags.json") as file:
    flags = json.load(file)
    for name in flags:
        get_country_by_name(name)["flag"] = flags[name]
    file.close()

for filename in sorted(os.listdir("resources")):
    if filename.endswith(".csv"):
        title = filename[:-4].replace("_", " ")
        if title not in report_titles:
            report_titles.append(title)
        with open("resources/" + filename) as file:
            next(file)
            rows = file.read().split("\n")
            rows.pop()
            for row in rows:
                tokens = row.split(",")
                country = get_country_by_name(tokens[1])
                if country != None:
                    country["reports"].append(
                        {"title": title, "rank": int(tokens[0]), "score": float(tokens[2])})
            file.close()

with open("resources/data.json", "w") as file:
    json.dump({"countries": countries, "reportTitles": report_titles}, file)
    file.close()
